### ISSUE TITLE
[C#-X.0 Proposal] Determinism

### ISSUE THREAD

**Issue by Unknown6656** [dotnet/csharplang#1763]
### EDIT: Jump to this comment -> https://github.com/dotnet/csharplang/issues/1763#issuecomment-410048484 <- to see the full proposal

------
# Purity / Determinism / Constant Functions / Constant Expressions

- [x] Proposed: https://github.com/Unknown6656/csharplang/blob/master/proposals/determinism.md
- [ ] Prototype: None
- [ ] Implementation: Not Started
- [ ] Specification: Not Started
------
I have written a rather long proposal about a possible language feature concerning the determinism/purity of code and it's usage for a future possible language version.

This could be seen as an "umbrella proposal" for the following issues (mixed with some of my ideas):

- [dotnet/csharplang#504](https://github.com/dotnet/csharplang/issues/504): "Enable Compile-Time Folding of Deterministic Outputs"
- [dotnet/csharplang#776](https://github.com/dotnet/csharplang/issues/776): "C# Pure Function Keyword to Mark No Side Effects Or External Dependencies" 
- [dotnet/csharplang#1028](https://github.com/dotnet/csharplang/issues/1028): "Compile time expressions"
- [dotnet/csharplang#1413](https://github.com/dotnet/csharplang/issues/1413): "Compile-Time method inlining"
- [dotnet/roslyn#9627](https://github.com/dotnet/roslyn/issues/9627): "Compile Time Function Execution"
- [dotnet/roslyn#10506](https://github.com/dotnet/roslyn/issues/10506): "Static Expressions"
- [dotnet/roslyn#11259](https://github.com/dotnet/roslyn/issues/11259): "Constant string interpolation"
- [dotnet/roslyn#12238](https://github.com/dotnet/roslyn/issues/12238): "Compile time expressions"
- [dotnet/roslyn#14665](https://github.com/dotnet/roslyn/issues/14665): "Constant expression - Power operator?"
- [dotnet/roslyn#15079](https://github.com/dotnet/roslyn/issues/15079): "How would you imagine `constexpr` in C#?"
- [dotnet/coreclr#3633](https://github.com/dotnet/coreclr/issues/3633): "Optimizing `constant_string.Length`"

---------

This is my first _in-depth_ proposal and I am open _(and hoping)_ for a healthy discussion about a subject which I myself find very interesting, potential-rich and very valuable for future C# iterations.

----------
Due to the length of the proposal, I would advise any reader to first grab a piece of cake and a hot chocolate/coffee before starting to read :wink:

**Comment by Unknown6656** [comment]
**/CC:** @gafter @MadsTorgersen 
@eyalsk (IIRC you did also lead a lengthy discussion about exactly this topic)

**Comment by gafter** [comment]
Community-submitted proposals should start as issues, not as PRs. Those issues can than gather community feedback. We use "proposal" PRs to track specifications for features that are already championed and approved (at least in principle).

**Comment by gafter** [comment]
To clarify: I recommend you place your actual proposal text in some issue (e.g. this one).

**Comment by Unknown6656** [comment]
@gafter: Do please excuse me -- I thought it was more practical to edit it locally and push it to my own fork instead.
I will copy the contents of the .md into this issue.

**Comment by Unknown6656** [comment]
_Original proposal:_
--------------
# Purity / Determinism / Constant Functions / Constant Expressions

This proposal is dedicated to the often mentioned ideas of _"function purity"_, _"determinism"_, _"pre-compilation"_ and _"constant functions/expressions"_.

This article also mentions features and wishes expressed (more or less briefly) the following proposals:

- [dotnet/csharplang#504](https://github.com/dotnet/csharplang/issues/504): "Enable Compile-Time Folding of Deterministic Outputs"
- [dotnet/csharplang#776](https://github.com/dotnet/csharplang/issues/776): "C# Pure Function Keyword to Mark No Side Effects Or External Dependencies" 
- [dotnet/csharplang#1028](https://github.com/dotnet/csharplang/issues/1028): "Compile time expressions"
- [dotnet/csharplang#1413](https://github.com/dotnet/csharplang/issues/1413): "Compile-Time method inlining"
- [dotnet/roslyn#9627](https://github.com/dotnet/roslyn/issues/9627): "Compile Time Function Execution"
- [dotnet/roslyn#10506](https://github.com/dotnet/roslyn/issues/10506): "Static Expressions"
- [dotnet/roslyn#11259](https://github.com/dotnet/roslyn/issues/11259): "Constant string interpolation"
- [dotnet/roslyn#12238](https://github.com/dotnet/roslyn/issues/12238): "Compile time expressions"
- [dotnet/roslyn#14665](https://github.com/dotnet/roslyn/issues/14665): "Constant expression - Power operator?"
- [dotnet/roslyn#15079](https://github.com/dotnet/roslyn/issues/15079): "How would you imagine `constexpr` in C#?"
- [dotnet/coreclr#3633](https://github.com/dotnet/coreclr/issues/3633): "Optimizing `constant_string.Length`"

  
# Introduction

C# has been and currently is moving into the exiting field of functional programming. It is not at all a new field _(think about how old Haskell is!)_ - but it is a field with great potential and should therefore be explored and used in modern C# programming.

One important function is the so-called **determinism ** of functions and expressions. One can gain extreme performance, parallelism improvements if rightly used and analyzed. Furthermore, deterministic compiler approaches can prevent many run-time exceptions due to improved code transformation and analysis.

I will use the word **determinism** throughout most of this article - but you can think of it as  *pure functions*, *pre-compiled functions* or *pseudo-"constant" functions*.


## What does 'Determinism' mean?

In a broad mathematical and functional sense, determinism is the concept of function or expression declaration which always returns the same output for the same input.<br/>Mathematically speaking, this should look as follows:

If any function ![](https://latex.codecogs.com/png.latex?%5Cinline%20%5Cdpi%7B100%7D%20%5Clarge%20f%20%3A%20%5Cmathcal%7BA%7D%20%5Cmapsto%20%5Cmathcal%7BB%7D) is deterministic, then follows:
![](https://latex.codecogs.com/png.latex?%5Cinline%20%5Cdpi%7B100%7D%20%5Clarge%20%5Cforall%20a_1%2C%20a_2%20%5Cin%20%5Cmathcal%7BA%7D%20%3A%20a_1%20%3D%20a_2%20%5CLeftrightarrow%20f%28a_1%29%20%3D%20f%28a_2%29)



A C# example for deterministic functions are, e.g.:

```csharp
public static long FuncA() => 420L;

public static long FuncB(int x)
{
    long res = 0;
    
    for (int c = 0; c < x; ++c)
        res += c * 42;
    
    return res;
}
```
It is pretty obvious, that the function above always outputs the same result for the same input. But what about the following one?

```csharp
public static long FuncC(int x)
{
    long input = long.Parse(System.Console.ReadLine());
    
    return input + x * 42;
}
```

As `FuncC` uses `Console.ReadLine : void -> string`, the function `FuncC` can only be deterministic, if `ReadLine` is (which it of course isn't).

This example shows that for the purpose of this article, one shall define 'deterministic as follows':<br/>**A function is deterministic ⟺**

1. All parameters are deterministic
2. All inner function calls are deterministic
3. Any expression, expression chain or operation 'flow' is deterministic
4. All dependencies are deterministic (in time)

_Constant functions (functions composed only of constant values and expressions) are trivially deterministic._



## How can we differentiate between deterministic and non-deterministic functions?

To determine whether a function is deterministic (aka. "pure"), one must (recursively) look at all affected functions and variables:

- If a function **solely** accesses parameters, (constant) literals, local function variables and other deterministic functions, then it **definitely** is deterministic.

- If a function accesses **any** of the following components - it is **not** deterministic: 

  - Memory (physical or virtual)
  - Interrupt-based mechanics, such as devices (including the mouse) or timers
  - Stream-based mechanics, such as Networking, I/O, etc.
  - UI-based components, such as Windows, UI-message-queues (they often use interrupts)
  - P/Invoke calls (or any calls to non-deterministic code)

  The above components are also called _"side-effects"_, as any code operating with them can have undesired or non-deterministic _(read: non-repeatable)_ side-effects on the own application's state.

- If a function access _global_ (static or instance) variables, the determinism cannot always be ensured. The requirements for this must be: No external or non-deterministic code can access the variables in question. All functions accessing the global variable must be deterministic.

- For any other remaining cases, one must define in greater detail where the line has to be drawn between determinism and non-determinism.



# Why do we need this?

Good point. Why do we need this indeed?<br/>There are a couple of reasons out there:

### 1.) Performance: caching or look-up

Imagine having having a huge dataset with statistical data. Determinism could be used in a lot of mathematical functions handling this dataset. Examples could be the calculation of metrics, such as the standard deviation, median, regression expressions, etc...

Many performance aspects however come apparent on the _second_ call of a deterministic function:<br/>As the JIT **knows** a function to be deterministic, it could simply look-up its result upon re-calculation. The result would be correct, as the output of **any** deterministic function does not change, as long as the input parameters  remain unchanged.<br/>The JIT could have some kind of (partial) look-up table or cache for deterministic functions in order to hugely improve an application's performance.

### 2.) Performance: parallelism

As deterministic functions are only composed of deterministic function-calls and expressions, their code is **known** to be **side-effect free**. They can therefore often be parallelized safely in order to gain performance, as they are known to have no side-effects on the remaining application state.

If global variables are involved, some kind of functional dependency graph or transactional system has to be created by the compiler in order to insure the variable's determinism.

You could imagine the parallelism of deterministic functions along the lines of parallelism of `async` functions:

Having a set of deterministic functions with function `F` waiting for the result of function `G` at some point, one can pre-compute `F` in the background (in parallel) and "await" its result before passing it along to `G`, e.g:

```csharp
// deterministic
public static const string F(string s)
{
    return new string(s.ToUpperInvariant()
                       .Reverse()
                       .ToArray());
}

// non-deterministic
public static void G(object o) => Console.WriteLine(o?.ToString() ?? "<null>");

// non-deterministic
public static void Main()
{
    string input = Console.ReadLine();
    
    Thread.Sleep(1000); // or some other long-duration operation
    
    G(F(input));
}
```

If the calculation times are as follows:

```
                 F:    5ms
  Console.ReadLine:   10ms [+ wait for user input]
 Console.WriteLine:   10ms
Thread.Sleep(1000): 1000ms
```

one would expect the e.g. following _classic_ timeline:

```
ms: 0                           5000   5010      6010 6015   6025
    |-----WAIT FOR USER INPUT-----|------|---------|----|------|
  START                           READLINE  SLEEP    F  WRITELINE
```

The method `F`, however, is deterministic, so it could already be called at `5010 ms` instead of `6010 ms`. This is semantically perfectly valid, as `F` is **known** to have **no side-effect** on the rest of the application (or the rest of the application on `F`).<br/>The execution time of `Main` could therefore be effectively reduced by the calculation time of `F` (5 ms):

```
ms: 0                           5000   5010 5015     6010   6020        <-- overall time reduced
    |-----WAIT FOR USER INPUT-----|------|----|--------|------|
    |                             |      |--F-'        |      |         <-- F is executed
  START                           READLINE    SLEEP    WRITELINE              in parallel     
```

_Of course - this was an example with rather bad numbers, but you do get the idea_... ;)

### 3.) Performance: pre-compilation

Imagine the following function `sigmoid`:

```csharp
using static System.Math;

public static float sigmoid(float x) => Exp(x) / (Exp(x) + 1);
```

Assuming that the function `Math.Exp : float -> float` is deterministic, the function `sigmoid ` obviously also is. Upon compiling a program which contains the function `sigmoid`, the compiler could take a look at all call occurrences and replace the function calls with the pre-calculated result.

This could also be applied to string-manipulation, e.g.:

```csharp
string s = $"Hello {Math.PI:N3}!".ToLowerInvariant();
```

would _always_ result in the following value:

```csharp
"hello 3.142!"
```

### 4.) Elimination of (semantically) unused code

As deterministic functions are per definition side-effect free, one can safely remove code snippets inside those functions which have no effect on the return value. This is possible, as the semantics of the "optimized" functions do not differ from the "original" one.

An example:

```csharp
L01: public const float MyComplexFunction(float y)
L02: {
L03:     float x = MyComplexFunction(Sin(y));
L04:     
L05:     return y * 7;
L06: }
```

Traditionally, the execution of `MyComplexFunction : float -> float` would result into a `StackOverflowException` on line `L03` due to a never-ending recursive call. This could not be optimized with the traditional C# compiler, as the function is not even tail-recursive.

However, as we know that `MyComplexFunction` is deterministic, it is known to have no side effects. Therefore, we can omit the line `L03` entirely, as its results will never effect the return value of `MyComplexFunction`.

The compiler could now remove `L03` **and** issue the hint `Removed line L03, as it has no effects on the function's semantics.`, instead of only issuing a warning `Unused variable 'x' in line L03.` and _leaving the line `L03` untouched._

The compiled result would be:

```csharp
public const float MyComplexFunction(float y) => y * 7;
```

Which would operate as **semantically** expected.

### 5.) Reduction of error sources

The usage of determinism could greatly reduce the amount of produced errors, as potential conflicts could be determined by the compiler (e.g. eventual overflow or division-by-zero). If the application of determinism in C# could be broadened to -- let's say -- discriminated unions and pattern matching, many patterns could be detected by the compiler which would e.g. never be matched etc.

Determinism could also be applied to the upcoming C#8.0-feature of nullable reference types: Determinism could provide the compiler with improved methods of tracking null references.

### 6.) More compile-time constants!!

Many expressions can be transformed to be compile-time constants, if all underlying (arithmetic) operators and function calls are insured to be deterministic. This enables the usage of compile-time constants in many more places.

Please take a look at the following chapter (especially [this section](#constant-expressions)) for more information.



# OK, I get it. But what would it look like?

There are multiple designs which this proposal could adopt:

- **"Conservative determinism":**<br/>The compiler should only employ determinism if the developer says to do so. This could implemented as follows:
  - A keyword like `const` or `pure` to mark deterministic functions.<br/>_My favorite solution for now is the usage of the existing keyword `const`, so I will stick to that throughout this article._
  - An compiler-intrinsic attribute, e.g. `[Pure] `or `[Determinsitic]` which would have the same effect as the keyword(s) mentioned above.
- **"Aggressive determinism:"**<br/>The compiler shall _greedily_ assume any function as being deterministic and require the developer to mark non-deterministic functions as such.<br/>This could be the proposal's "final" goal in the far-away future. However, such a behavior would be a breaking change at the current stage.

For the purpose of this proposal, we settle on **"Conservative determinism"** with the keyword `const` as determinism-marker for now.

-----

## Examples:

### Deterministic functions

```csharp
/// calculate hyperbolic cotangent
public static const float Coth(float x) => (Exp(2 * x) + 1) / (Exp(2 * x) - 1)
```

This would the function `System.Math.Exp : float -> float` also require to be marked as `const`.<br/>Any call **_from_ a deterministic** function **_to_ a _non_-deterministic** one would result in a compiler **error**:

```csharp
public static DateTime GetToday() => DateTime.Now;

public static const DateTime GetNextDay()
{
    DateTime today = GetToday();
    //               ^^^^^^^^^^ ERROR
    // A non-deterministic function cannot be called from a deterministic one.
    
    return today.AddDays(1);
}
```

If the function `GetToday : void -> DateTime` would be marked as `const`, the error would "shift" as follows:

```csharp
public static const DateTime GetToday() => DateTime.Now;
//                                                  ^^^ ERROR
// A non-determinisitic property cannot be accessed from a determinisitic function.     

public static const DateTime GetNextDay()
{
    DateTime today = GetToday();
    
    return today.AddDays(1);
}
```

It is needless to say that calls **_from_ a _non_-deterministic** function **_to_ a deterministic** one are perfectly **valid**.

### Deterministic types

A type can be marked as deterministic, if it is `sealed` and all containing methods are deterministic. Furthermore, deterministic types cannot be modified via memory operations such as pointer, pinning or P/Invoke operations:

```csharp
public const struct Point2Df
{
    public const float X { get; }
    public const float Y { get; }
    
    ....
}


Point2Df point = new Point2Df();
Point2Df* ptr = &point;
//      ^       ^^^^^^ ERROR
// Cannot take address of deterministic datatyoe 'Point2Df'.
```

For now, only classes and (managed) read-only structures should be valid candidates for deterministic datatypes.

A deterministic type could look as follows:

```csharp
public const sealed class ComplexNumber
{
    private float _re, _im;
    
    
    // deterministic constructor
    public const ComplexNumber()
        : this(0, 0) { }
    
    // deterministic constructor
    public const ComplexNumber(float re)
        : this(re, 0) { }
    
    // deterministic constructor
    public const ComplexNumber(float re, float im)
    {
        _re = re;
        _im = im;
    }
    
    // deterministic function (this requires "Sqrt : float -> float" to be deterministic)
    public const float GetMagnitude() => Sqrt(_re * _re + _im * _im);
}
```

### Deterministic operators and properties

Properties and (custom) operators can only be used inside deterministic functions or expressions, if they are also marked as `const`. The syntax would look along these lines:

```csharp
public const sealed class ComplexNumber
{
    .....
    
    // deterministic read-only property
    public const float Imaginary => _im;
    
    // deterministic read-write property
    public const float Real
    {
        get => _re;
        set => _re = value;
    }
    
    public static const ComplexNumber operator+(ComplexNumber c1, ComplexNumber c2) =>
        new ComplexNumber(c1._re + c2._re, c1._im + _c2.im);
}
```

The code above could be compiled into something like this:

```csharp
public const sealed class ComplexNumber
{
    .....
    
    public const float get_Imaginary() => this._im;
    
    public const float get_Real() => this._re;
    
    public const void set_Real(float $value) => this._re = $value;
    
    public static const ComplexNumber op_Addition(ComplexNumber c1, ComplexNumber c2) =>
        new ComplexNumber(c1._re + c2._re, c1._im + _c2.im);
}
```

which are all deterministic functions.

### Deterministic flow control

Determinism would also expand to flow control, e.g.:

```csharp
public static const float GetMagicNumber(float x) => x / 0f;
// Compiler detects that the function above always returns float.Infinity

public static const string GetMagicString(int i)
{
    float res = GetMagicNumber(i + 42.0f);
    
    if (float.IsInfinity(res))
        return "Oh noes!";
    else
    {
        return "Yes!";
    }
}
// As all used constructs are deterministic, the compiler would replace any call of the
// function 'GetMagicString : int -> string' with the constant string "Oh noes!".
```

### Constant expressions

Determinism naturally extends to expressions, meaning that the expression `420f / 10f` is expected to always yield the result `42f`. One could therefore redefine compile-time constant expressions:

**The value of an expression is known at compile-time if it is only composed constant literals and deterministic function calls with constant parameters.**

This would e.g. enable the following expression to be compile-time constant:

```csharp
public const int MY_CONSTANT = $"Hello, {Math.E}!"[7] - '\x8';
```

It would be composed as follows during compile-time:

```csharp
public const int MY_CONST = $"Hello, {Math.E}!"[7] - '\x8';
//  equals:                 (int)string.Format("Hello, {0}", Math.E).get_Chars(7) - (int)'\x8'
//                          (int)     "Hello, 2.71828182845905!"    .get_Chars(7) - (int)'\x8'
//                          (int)                       '2'                       - (int)'\x8'
//                                      int.op_Implicit('2')      -      int.op_Implicit('\x8')
//                                             int.op_Subtract(50, 8)
//                                                       42
```

The compiler would build an expression tree of the expression `MY_CONSTANT : int` and only evaluate its value if the following functions are deterministic or constant:

```
string.Format : string -> obj[] -> string
string.get_Chars : string -> int -> char
int.op_Implicit : char -> int
int.op_Subtract : int -> int -> int
```

The compiler would then issue the following IL for the expression above:

```assembly
.field public static literal int32 MY_CONSTANT = int32(42)
```



Constant expressions could be used as default parameter value:

```csharp
public static float FuncF(float x, float y = Sin(2 * PI) / 14)
//                                           ^^^^^^^^^^^^^^^^
//                                           This will be evaluated at compile-time!!
{
    return Tan(x) * y;
}
```



Naturally, the usage of compile-time constants (in the deterministic sense) could be expanded to attributes:

```csharp
[Obsolete($"Perfectly valid code! {nameof(MyFunc)} is obsolete -- and this is compile-time constant! {Math.Pow(0.025, -1) + 2}")]
public void MyFunc() { ... }
```

Which would evaluate to:

```csharp
[Obsolete("Perfectly valid code! MyFunc is obsolete -- and this is compile-time constant! 42")]
public void MyFunc() { ... }
```

### Deterministic Attributes (?)

A possible idea would be to allow determinism to be allowed in attributes as well:

```csharp
public const delegate bool DeterministicFunc(float f);

public const sealed class CheckParameterAttribute
    : Attribute
{
    ...
    
    const public CheckParameterAttribute(DeterministicFunc<float, bool> predicate)
    {
      ...
    }
}


public const void funcE(
    [CheckParameter(const f => f > -1 && f < 1)]
    float f
)
{
  ...
}
```

This would, however, require a language change (except if a deterministic function handle or `.text` offset would be stored as a constant inside the attribute ....)

### What about accessing type/global variables?

Well -- this is a more complex matter. Imagine having the following code:

```csharp
public const struct Matrix2D<const T>
{
    private T[,] _values;
    
    
    public const Matrix2D(int n, int m) =>
        _values = new T[n, m];
    
    public const void Transpose() { ... }
    public const T[] Eigenvalues() { ... }
    public const Vector<T> Eigenvectors() { ... }
    public const (Matrix2D<T> Left, Matrix2D<T> Right) DecomposeLR() { ... }
    
    public static const Matrix2D<T> GenerateTHEmatrix() { ... }
}
```

Let's imagine the following access patterns:

| method                                            | variable             | access       |
| ------------------------------------------------- | -------------------- | ------------ |
| `.ctor : int -> int -> Matrix2D<T>`               | `_values` (instance) | write        |
| `Transpose : void -> void`                        | `_values` (instance) | read, write  |
| `Eigenvalues : void -> T[]`                       | `_values` (instance) | read         |
| `Eigenvectors : void -> Vector<T>`                | `_values` (instance) | read         |
| `DecomposeLR : void -> Matrix2D<T> * Matrix2D<T>` | `_values` (instance) | read         |
| `GenerateTHEmatrix : void -> Matrix2D<T>`         | &lt;none&gt;         | &lt;none&gt; |



If the code of our `Main`-function looks as follows:

```csharp
public static const void Main(string[] argv)
{
    Matrix2D<float> l, r, m, n;
    Vector<float> ev;
    float[] e;
    
    m = new Matrix2D<float>(5, 5);
    .... // fill values into m
        
    n = Matrix2D.GenerateTHEmatrix(); // some really time-intensive operation
        
    e = m.Eigenvalues();
    ev = m.Eigenvectors();
        
    m.Transpose();
    
    (l, r) = m.DecomposeLR();
}
```

One cannot parallelize `Transpose` and `DecomposeLR` trivially, as they have conflicting access patterns to the same type variable (in this case `m._values : float[]`). However, the function calls of `Eigenvalues` and `Eigenvectors` are parallelizable, as both only require a read-access on the type variable `m._vlaues`.

The compiler must create some kind of functional dependency graph (a bit like transactional database systems do), in order to parallelize as much code as possible:

<img src="https://raw.githubusercontent.com/Unknown6656/csharplang/master/proposals/img1.png"/>

 _[Ignore the absolutely mad MSPaint skillz™]_<br/>On the left-hand side one can see the serial (imperative) execution of the functions in the code example above. After building a deterministic execution tree, the compiler would use the dependency graph on the right-hand side in order to parallelize as much code as possible without causing semantic inconsistencies.

The following access conflict table should be used when checking whether two sequential accesses `T1` and `T2` can be parallelized when accessing the same variable `x`:

|     `T1 -> T2`      | `T1` : read(x) | `T1` : write(x) |
| :-----------------: | :------------: | :-------------: |
| **`T2` : read(x)**  |      yes       |       no        |
| **`T2` : write(x)** |       no       |       no        |

If the dependency graph shows a cyclic dependency, all deterministic functions must be executed in sequence (traditional code execution order).

----

### TL;DR

In general, the compiler could follow the following guidelines for deterministic functions:

- no parameters, no type variable access ⟹ calculate function result **at compile-time**.
- parameters, but all calls pass constant values ⟹ calculate function result **at compile-time**.
- parameters, no type variable access ⟹ calls can be calculated **asynchronously**.
- type variable access ⟹ compiler has to resort to a dependency graph etc. in order to insure correctness.

# Issues / Drawbacks / Questions

This goes without saying, that such complex determinism analysis requires a lot of work. The following points are some issues, drawbacks and open questions which I thought of:

- **This requires a language change.** The `const`- or `[Deterministic]`-marker could be issued as IL-Metadata, but I am not sure whether it is enough. IMO, `const` should be part of a deterministic function's signature -- however -- functions should not be overloadable by the `const`-marker only.<br/>A language change is definitely required if determinism is to be allowed for Attribute constructors with non-constant _(but deterministic)_ attributes.

- **Core framework changes.** Many, many CoreFX functions could be transformed to be deterministic, e.g. most mathematic and LINQ functions.

- **How are deterministic APIs exposed?** That's a very good question.

  - **Should they act like regular functions?** That would detain the compiler from pre-compiling some deterministic functions.
  - **Should precompiled functions be accessible as constants?** Maybe - but that does not sound like .NET-consistent compiler behavior.

- **What happens with reflection-invoked deterministic functions?** They should act like API-exposed ones; meaning that the solution to the API-question above would be the same for reflection-invocation.

- **Deterministic Generics? Interfaces?** Definitely!<br/>One could introduce some constraints as follows:

  ```csharp
  public interface IVector<const T>
  {
      const T ScalarZero { get; }
      const T ScalarOne { get; }
      
      const IVector<T> VectorZero { get; }
      const IVector<T> VectorOne { get; }
      
      const IVector<T> VectorAdd(IVector<T> other);
      const IVector<T> ScalarMultiply(T scalar);
  }
  ```

- **What about deterministic λs and delegates?** Maybe like this:

  ```csharp
  const delegate bool DeterministicPredicate<const T>(T value);
  
  
  public const T[] LINQWhere<const T>(T[] input, DeterministicPredicate<T> predicate)
  {
      List<T> result = new List<T>();
      
      foreach (T element in input ?? new T[0])
          if (predicate(element))
              result.Add(element);
      
      return element.ToArray();
  }
  
  public static const void Main(string[] argv)
  {
      int[] source = { 3, 1, 5, 4, 2, -88 };
      int[] result = LINQWhere(source, const i => (i % 2) != 0);
  }
  ```

# Conclusion

**IS THIS WORTH IT?**<br/>Yes - I **definitely** think that the effort is worth the drawbacks for a long-time future.

The C#8.0 feature `nullable reference types` already goes into the right direction. However, I think that a whole lot more potential and performance can be gained from this proposed language feature.

IMO, It would definitely require a breaking CLI change (which would incidentally be a good excuse to clean-up the CLI and its instruction set :wink:).

The huge workload involved in the implementation would therefore make it a candidate for a major language version (e.g. C#-10).

**Comment by HaloFour** [comment]
> IMO, It would definitely require a breaking CLI change (which would incidentally be a good excuse to clean-up the CLI and its instruction set 😉).

This pretty much guarantees that it's destined to go absolutely nowhere.

Giant meta-proposals are not useful.  They force all of the feedback through a single funnel making the following of any conversation completely impossible.  If you wanted any serious attention to any of these individual proposals I'd suggest submitting them separately, at least for the proposals that you're not just duplicating.

**Comment by svick** [comment]
Apologies in advance if some of my notes that follow get too nitpicky.

---

The term "determinism" refers to [quite a different concept in the Roslyn compiler](https://github.com/dotnet/roslyn/blob/master/docs/compilers/Deterministic%20Inputs.md), so I think this proposal should use a different name (like "pure functions"), to avoid confusion.

---

> All parameters are deterministic

What does that mean? It seems to me you are conflating whether a function (e.g. `F(int i)`) is deterministic and whether a function *invocation* (e.g. `F(42)`) is deterministic .

---

> Any expression, expression chain or operation 'flow' is deterministic

What are "expression chains" and "operation flows"? Isn't it enough to say that all statements need to be deterministic? (And that a statement is deterministic if all its child expressions and statements are deterministic.)

---

> All dependencies are deterministic (in time)

I don't understand what this means. What are a function's dependencies? Is this referring to field accesses?

---

> If a function accesses any of the following components - it is not deterministic:
>*    Memory (physical or virtual)
>*    Interrupt-based mechanics, such as devices (including the mouse) or timers
>*    Stream-based mechanics, such as Networking, I/O, etc.
>*    UI-based components, such as Windows, UI-message-queues (they often use interrupts)
>*    P/Invoke calls (or any calls to non-deterministic code)

What does it mean to "access memory" from C#? Does reading a local variable (which could be stored on the stack, which is part of memory) count? Does e.g. `"foo"[1]` count?

Also, don't most of these boil down to PInvoke calls anyway?

---

> Performance: caching or look-up

Memoization can indeed improve performance significantly. But I would like to see evidence that some form of automatic memoization at the JIT level would make sense. As far as I know, [even Haskell does not use that](https://stackoverflow.com/a/12390953/41071).

---

> Performance: parallelism

Your example would require adding some synchronization. How would the compiler figure out that doing so is worth it? Also, it's improving latency at the cost of decreasing throughput (e.g. when there are many requests and you want to process as many requests as possible, you don't want the computation to be done in parallel). How does the compiler determine that this trade-off is acceptable to the user?

In other words: automatic parallelization is a hard problem. Knowing what code can be safely executed in parallel is necessary for that, but not sufficient.

---

> This could also be applied to string-manipulation, e.g.: `string s = $"Hello {Math.PI:N3}!".ToLowerInvariant();` would always result in the following value: `"hello 3.142!"`

That would be a breaking change, since it currently produces `"hello 3,142!"` on my machine.

---

> Elimination of (semantically) unused code

Do you have a more realistic example of where this would be actually useful?

---

> Reduction of error sources

Do I understand it correctly that this is about producing errors at compile time for code that *always throws*? If not, could you expand on that?

---

> This would e.g. enable the following expression to be compile-time constant:

No, it would not. `string.Format` is not deterministic, because it depends on the current culture.

---

> The compiler must create some kind of functional dependency graph

How would the compiler do that? How would it figure out what fields does a function read or write? And if it can do that, wouldn't it also be able to figure out what functions are deterministic, so we wouldn't need the `const` modifier?

**Comment by eyalsk** [comment]
@Unknown6656 

Like @HaloFour I really think that you need to break this into multiple proposals and I'm not sure where you should start but it seems like touch multiple places, personally, I want to have constant functions and expressions.

I'll throw it out here because it's a **thought** I had from the previous discussion we had instead of letting the compiler do all this work I'd really think it's better to have a deterministic engine where you can mark any expression/function with `const` (or another keyword) and then spawned Roslyn as a service (which it is) to compute the value and replace it at the callsites or at the position of the expression during compilation, it also means you can go a bit out of the pure purity area and have things that touches the system just work but really this approach doesn't even need language changes, CLI changes or CLR changes, as a prototype you can fork Roslyn, create your own `ConstAttribute`, attribute and hook into the compiler to do all this work I don't have a lot of details but this is certainly doable, you can certainly have your own keyword as opposed to an attribute as you're forking it but as a prototype it might be easier to just use an attribute at first and focus on the engine that makes all this possible.

Good luck! ;)

**Comment by Unknown6656** [comment]
@svick 
> The term "determinism" refers to quite a different concept in the Roslyn compiler, so I think this proposal should use a different name (like "pure functions"), to avoid confusion.

Yes -- you are right, I should rename it

----------------
> All parameters are deterministic

That was a bad choice of wording on my part ... I meant that the arguments can be deterministically computed (at _invocation_)

----------------
> What are "expression chains" and "operation flows"? Isn't it enough to say that all statements need to be deterministic? (And that a statement is deterministic if all its child expressions and statements are deterministic.)

You are right, saying that all statements need to be deterministic is enough. I thought that I should elaborate this a bit, as devs sometimes tend to forget, that operators can also be functions (instead of simple CPU instructions).

----------------
> I don't understand what this means. What are a function's dependencies? Is this referring to field accesses?

Yes, but not only field access. I view a function's dependencies as a set of data sources, which a function relies upon. This could be access to a variable/field/... at a point `X` in time. It could also be a sort of execution "contract", that some other code is executed e.g. before the time-point `X`. Temporal determinism would be a kind of assurance, that e.g. function `A` will _always_ be executed before function `B`.

----------------
> What does it mean to "access memory" from C#?

I meant "traditional" memory access via e.g. pointers, fixed buffers etc.. I mean, that determinism can only be assured, if no other code can modify the memory region utilised by the deterministic function.

> Does e.g. "foo"[1] count?

No. Nor does array access (as long as the array is not pinned).

> Also, don't most of these boil down to PInvoke calls anyway?

Do you mean the points in my list? Yes, they more or less do.

----------
Concerning your thoughts on `Performance: parallelism`: Hm ... I see your points. One must maybe resort to compiler hints via Attributes. Now you mention it, the problem is AFAIK indeed NP-Hard. That could be pretty difficult for a large code base to compile....

-----------
> That would be a breaking change, since it currently produces "hello 3,142!" on my machine

Hm -- interesting. I forgot that some cultures differentiate between `,`  and `.` as decimal separator.

------------
> Do I understand it correctly that this is about producing errors at compile time for code that always throws?

Yes - exactly.

------------
> No, it would not. string.Format is not deterministic, because it depends on the current culture.

I unfortunately forgot that during the typing of the proposal - I never really used culture-dependent string formatting (not even when designing language packages for applications).

-----------
> How would the compiler do that? How would it figure out what fields does a function read or write? 

I think that this should be possible by looking at the generated syntax tree and symbol table to resolve all assignments to/reads from fields.
> And if it can do that, wouldn't it also be able to figure out what functions are deterministic, so we wouldn't need the `const` modifier?

Yes -- you are right. It would effectively render the `const` modifier obsolete (which would kind of be the final goal: the compiler detects determinism automatically)

**Comment by Unknown6656** [comment]
@eyalsk @HaloFour 
You are right -- I should maybe break up this proposal into multiple issues (as long as they have not already been proposed).

@eyalsk I like your idea of Roslyn acting as a service in the background.
I would also be very interested in creating such a Roslyn fork and trying to develop a prototype. The problem, however, is that I currently do not have enough time for such a project (mainly due to university and work). If I find enough time, I will dig deeper into this!

**Comment by jnm2** [comment]
@svick 

> And if it can do that, wouldn't it also be able to figure out what functions are deterministic, so we wouldn't need the `const` modifier?

To some extent it's important to use a modifier, especially if you want determinism to act across binary boundaries. I see it as similar to the proposed `readonly` keyword they're working on for methods right now (there's a better example I'm not thinking of), or the principle that the return type should not be inferred because of spooky action at a distance.

**Comment by svick** [comment]
@jnm2 

> To some extent it's important to use a modifier, especially if you want determinism to act across binary boundaries.

That's part of my question. As I understand it, this proposal would require knowing which fields are read or written by a `const` method, even across assembly boundaries. How does that work? Since the C# compiler inspecting IL is almost certainly not realistic, do you have some set of attributes that would expose what until now were implementation details? (Which I think would also cause that "spooky action at a distance".)

**Comment by orthoxerox** [comment]
> That was a bad choice of wording on my part ... I meant that the arguments can be deterministically computed (at invocation)

If that argument is a delegate, can your function be inferred to be deterministic if that delegate is? Like, `Enumerable.Select(this source, func)` is deterministic if `func` is?

**Comment by Unknown6656** [comment]
@orthoxerox : That was my thought exactly.

**Comment by jrmoreno1** [comment]
@Unknown6656: @eyalsk idea seems a lot better than spending time on ways to determine if a function is deterministic.  Naive approach, create a flag for calculated consts, use that to compile 2 versions of the code one without and one with, use the one without to supply values to the one with.  Doubles the work that needs to be done, but everything that can be calculated and used could be used, and the only restriction is that all of the direct function calls be static

**Comment by thargol1** [comment]
I made an example where it's up to the programmer to determine if a function is deterministic:

```csharp
[Deterministic]
public static int Square(int i)
{
	return i * i;
}

[Deterministic]
public static string HtmlEncode(this string value) =>
	value == null ? string.Empty : HttpUtility.HtmlEncode(value);

[Deterministic]
public static string RemoveIndent(this string value)
{
	// implementation ommited
}

[Deterministic(omitFromFinalCode : true)]
public static string ReadFileAsText(this string filename)
{
	// implementation ommited
}

[Deterministic]
public static string BuildDateTime() => DateTime.Now.ToString("yyyyMMddhhmmss");

const int i = Square(3); 
// Compiles to const int i = 9;

const string s = "<b>test</b>".HtmlEncode(); 
// Compiles to const string s = "&lt;b&gt;test&lt;/b&gt;";

const string template = 
	@"<div>
		<span class='great'>C#</span>
	</div>".RemoveIndent();
// compiles to const string template = @"<div><span class='great'>C#</span></div>";

const string template2 = ReadFileAsText("template.html").RemoveIndent();

const string CacheBuster = BuildDateTime();
```
See also #2176, as this might provide a solution for the indented string literals.

This does not require a language change but a compiler change and perhaps, if all agree, this should be moved to the roslyn section.

A lot of the framework and core functions are deterministic and would have to be marked as such e.g. `string.ToUpperVariant()` or `Math.Floor()`.

And perhaps a parameter like `omitFromFinalCode` can be used to indicate that the function itself can be removed from the final output.

I also provided a non deterministic example above. The `BuildDateTime()` is actually non deterministic and shows how to this feature could be used for other purposes.

**Comment by ewlloyd** [comment]
While automatic determinism (through syntax tree analysis) is a laudable goal, I do like the idea of having the syntactic salt of a `const` or `pure` keyword available to declare intent. That way, the compiler can catch you out if you accidentally introduce a non-deterministic element to your code.

**Comment by deinok** [comment]
@Unknown6656 Is this a port to C# of the `constexpr` from C++ 11?

**Comment by Joe4evr** [comment]
@deinok No, this feature is purely to guarantee a compiler semantic of "exact same source code (and compiler options) goes in" means "exact same IL binary comes out" regardless of the system environment that is doing the compilation.

**Comment by Unknown6656** [comment]
@Joe4evr I would partly like to disagree on that ..... my focus in this issue is not on compiler determinism, but rather on execution determinism -- so it is in a sense related to `constexpr` mentioned by @deinok.

**Comment by deinok** [comment]
@Unknown6656 Here https://en.cppreference.com/w/cpp/language/constexpr you can see some rules required for something to be `constexpr` valid. They mostly match your proposal. 

Also, could you explain why this requires a CLI breaking change? I think it important to know why

**Comment by Unknown6656** [comment]
At the time I wrote the proposal I thought that a breaking change may be necessary to greedily enforce determinism (the compiler shall assume any function possible to be `constexpr`/`pure`/...), however, I do think that it may be possible without breaking changes if this could be introduced as a comiler-optin (like nullable refrence types are)

**Comment by deinok** [comment]
I think that this case if we don't take the greedily enforcement won't require a change in the CLR as it works only as a optimization at the generation of the resulting MSIL level

**Comment by scalablecory** [comment]
`constexpr` being done in C# makes sense if we look at how `const` is handled today.

I would expect the JIT to automatically fold constants for non-constexpr when possible... maybe the `AggressiveOptimizations` hint might tell it to do so.

**Comment by Unknown6656** [comment]
@deinok I concur.
@scalablecory I think that we could (to a certain degree) go beyond constant folding when possible, e.g.:
```csharp
public static int calculate_stuff(int a, int b)
{
    while (b > 0)
    {
        a ^= a << 1;
        --b;
    }

    return a;
}

int x = calculate_stuff(1, 5);
int y = calculate_stuff(x % 10, 3);
```
should be replaced by the compiler with
```csharp
int x = 51;
int y = 15;
```
as all inputs and calculations are completely deterministic.

As you  said that could be implemented with some kind of "agressive optimization"-switch (maybe along the lines of `[MethodImpl(MethodImplOptions.AgressiveInlining)]` combined with some more sophisticated constant folding....).

**Comment by yaakov-h** [comment]
I believe the JIT can do some of that already, even though the compiler cannot.

**Comment by CyrusNajmabadi** [comment]
> should be replaced by the compiler with

This requires the compiler to have the ability to evaluate arbitrary code.  This also raises serious concerns around things central to the design of roslyn (like fast compilation).  

I would far prefer this sort of optimization be done as an offline pass on the IL rather than be part of the language.  This way:

1. the language spec doesn't get even more bloated by having to write out the entire semantics of all of this.
2. you get the benefit for *all* .net languages, not just c#.
3. you pay the perf hit once, instead of on every compile.

**Comment by deinok** [comment]
@CyrusNajmabadi When you say "offline pass on the IL" you mean the JIT converting the MSIL to code executable by the CPU?

**Comment by jnm2** [comment]
@deinok JIT wouldn't be called offline. Cyrus is talking about a new post-build IL weaving tool.

**Comment by deinok** [comment]
@jnm2 Okey, point 3 makes me feel like he was talking about something at runtime insted of compile time

**Comment by CyrusNajmabadi** [comment]
> @jnm2 Okey, point 3 makes me feel like he was talking about something at runtime insted of compile time

I meant 'compile time'.  However, you only do it one final time before deploying, instead of every time csc compiles.   This is basically an IL optimizer driven by attributes to indicate what's acceptable to precompute.

**Comment by Unknown6656** [comment]
@CyrusNajmabadi : Your variation of the proposal would be perfect and I do agree with you on all points.
I would like to see this kind of "determinism optimiser" as tool which can be appended to the compiler pipeline for any .NET language (this could either be a compiler opt-in or a VisualStudio background service which updates all (re-)compiled assemblies non-blocking in the background).

Do you know how one could get a proposal for such a tool or pipeline-step started?

**Comment by Joe4evr** [comment]
@Unknown6656 See dotnet/roslyn#15929.

Additionally, the [IL Linker](https://github.com/dotnet/core/blob/master/samples/linker-instructions.md) tool already exists for the purpose of producing a single .exe and removing dead code, but it's very possible that it could get extended or something else gets based on it to do more granular optimizations.

**Comment by Unknown6656** [comment]
@Joe4evr oh that's good to know ... I will take a closer look at the IL Linker.
@CyrusNajmabadi  I have thought about one issue, however: what about Reflection?
I have that opinion that a compiler should aggressively optimize anything that is deterministic, however this will be an issue when one wants to invoke a function which would otherwise be optimized away, e.g.:
```csharp
private static int do_stuff(int a, int b) => a < b ? a + b : a - b;

public static int Main(string[] argv)
{
    int a = 42;
    int b = 2;

    return do_stuff(a, b);
}
```
which should be compiled to the following when enforcing determinism optimizations:
```csharp
// 'do_stuff' should not be emitted as it not publicly available and is completely deterministic
// but now we cannot access it via reflection.

public static int Main(string[] _)
{
    return 40;
}
````
Shouldn't we need some kind of indicator which method can be completely optimized away (keyword, pseudo-attribute, ....)?

**Comment by theunrepentantgeek** [comment]
If you want to completely optimize away any (every?) pure/deterministic function that has parameters known at compile time, the compiler is going to need to _execute_ the code to generate the result.

Given that the Roslyn compiler is extensively used within Visual Studio (for Intellisense, Unit Test discovery, Code Analyzers, other things), this raises a scary scenario:

I clone a git repo of something, so I can review the code.
Unknown to me, the repo has been compromised by a bad actor.
I load the code into Visual Studio for review.
Visual Studio uses Roslyn to evaluate IntelliSense, the malicious code exploits a bug and gets to run on my system.
Bad code runs on my machine - it's not my machine anymore.

Notes ...

... *I* didn't compile and run the code, it happened automatically in the background.
... the very tools that I'd be using to review the code to see if it was safe are the exact tools that could be compromised to exploit my system.
... there are always bugs, no matter how diligent or disciplined the team involved. From a security perspective, _assume breach_ and evaluate how much damage can be done.

Aggressive inlining of pure/deterministic functions might be worth it - but would clearly need to be approached in a very conservative, diligent and security conscious fashion.

**Comment by YairHalberstadt** [comment]
@theunrepentantgeek
TBH if you download a repo with compromised code you are almost certainly stuffed anyway. They could easily hide something in the code which you want notice, and does something malicious when executed.

Also I imagine it would be doable to only evaluate constant expressions when actually compiling, and/or to evaluate them in a sandbox. After if they are truly constant, they shouldn't require anything outside the sandbox.

**Comment by CyrusNajmabadi** [comment]
> I have that opinion that a compiler should aggressively optimize anything that is deterministic, however this will be an issue when one wants to invoke a function which would otherwise be optimized away, e.g.:

i would never expect the *actual* functions to be optimized away.  I would simply expect the *invocations* of those functions to be potentially optimized away.

**Comment by CyrusNajmabadi** [comment]
>Also I imagine it would be doable to only evaluate constant expressions when actually compiling, and/or to evaluate them in a sandbox. After all, if they are truly constant, they shouldn't require anything outside the sandbox.

If you're willing to wait till compilation time, then i go back to my original point: i don't see why this is part of the C# language.  Just add a compile step that does this through IL optimization.

**Comment by YairHalberstadt** [comment]
@cyrusnajmabadi
I don't necessarily disagree. I'm just addressing @theunrepentantgeeks security concerns.

**Comment by jnm2** [comment]
Arbitrary user code in the form of MSBuild scripts executes when you load a csproj. It can execute any code that a standalone EXE can. The same is true of each NuGet package you reference when you do a package restore.

**Comment by jnm2** [comment]
My advice is to be [antifragile](https://en.wikipedia.org/wiki/Antifragile) to malware. This means offline backups, only allowing your machine to access data that you don't mind sharing with the world, and a streamlined routine so that doing clean installations of Windows doesn't hurt.

**Comment by Unknown6656** [comment]
@theunrepentantgeek
Asides from "being careful" I would agree that the compiler shall execute the code in order to optimize it, **if** is has been marked as 'pure' (by the user or the compiler itself).

Think about it, what could it execute? Obviously it would not touch any machine-specific code, nor any code concerning streams, I/O, shared memory, networking, user input or hardware. These functions/APIs are not pure in a deterministic sense.
The compiler would only execute code concerning constant manipulation (numeric or string-wise), stuff like known collection operations etc.

I do not see any problem as any "harmful" code would obviously require to be harmful outside of the program itself in order to harm the machine (or any in/outgoing communication). As these are per definition non-deterministic, the compiler would not even touch them.

_(I'm using the 'compiler' here as step, but it could technically be in the linker or an other pipeline step)_

-------

@CyrusNajmabadi: Why should we not optimize away complete functions? Would it make a semantic difference on non-public functions (or fields for that matter)?
Obviously this would break reflection but I do not see any issues aside from that..... (well except that it would be a breaking change from the specification's point of view.)

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi: Why should we not optimize away complete functions?

What is a 'complete function'?

> Obviously this would break reflection

That's a pretty big deal.  Why would you break reflection?

> well except that it would be a breaking change from the specification's point of view

That also seems like a pretty big deal.

**Comment by Unknown6656** [comment]
The point I would like to make is:
If we want to optimize the program as much as possible why leave functions which could be eliminated and do not have any _public_ semantics?
One could alternatively try to inline the calls and leave the original functions (more or less) unchanged in the assembly so as to not break reflection.... but should we?

I do understand, that it is a huge deal, however I think that we should reflect upon about how we could reduce a .NET application's footprint and resource load.

**Comment by deinok** [comment]
@Unknown6656 I think that you are right, but this kind of optimization will be problematic, I mean not imposible without breaking-changes, but really dificult. I think that for now we should try to push the pure/determinism/constexp part of the proposal.

PD: Of course its good to have in consideration future work related to this proposal. This probably will help upstreaming this proposal

**Comment by CyrusNajmabadi** [comment]
>If we want to optimize the program as much as possible why leave functions which could be eliminated and do not have any _public_ semantics?

What do you mean by 'public' semantics?  reflection, for example, is a public API.  

Note: if you want something that pulls out unused methods, then just use a tree-shaking tool.  Note that none of this is really related to the language per-se.  It seems to be about optimizing the end compiled IL.  THis all feels like the domain of IL optimizers.

**Comment by Unknown6656** [comment]
> It seems to be about optimizing the end compiled IL. THis all feels like the domain of IL optimizers.

Yes it is mostly on an IL level and not language level, however, I thought some language annotation might be needed in my original proposal.

> What do you mean by 'public' semantics? 

Private methods not having any visible effect on the program's input and output

> reflection, for example, is a public API.

That is the reason why it is such a breaking change and why I "retract" my previous standpoint.

-----
> I think that you are right, but this kind of optimization will be problematic, I mean not imposible without breaking-changes, but really dificult. I think that for now we should try to push the pure/determinism/constexp part of the proposal.

I agree that this will be difficult. I would like to help on that matter but I do not know where would be the best point to start...

**Comment by deinok** [comment]
I think that the first step is view the benefits and harms of IL Level implementation vs Language Level implementation. But probably @gafter can give us more information of what is required and his personal opinion on this.
